### PR TITLE
fix(lambda-nodejs): esbuild detection with Yarn 2 in PnP mode

### LINKED ISSUE #13179: (aws-lambda-nodejs): does not properly detect `esbuild` when used with Yarn 2+ in PnP mode
`aws-cdk` in general works properly with yarn 2+ in Plug'n'Play (PnP) mode. See https://yarnpkg.com/features/pnp

PnP mode does not use `node_modules` and thus `npx` cannot detect that `esbuild` is installed in the project, so this block of code fails:

https://github.com/aws/aws-cdk/blob/8f1cfe4087b765cecd3ceb01745167646d617fec/packages/%40aws-cdk/aws-lambda-nodejs/lib/util.ts#L115-L130

Also this block of code is problematic as well since it also uses `npx`:
https://github.com/aws/aws-cdk/blob/8f1cfe4087b765cecd3ceb01745167646d617fec/packages/%40aws-cdk/aws-lambda-nodejs/lib/bundling.ts#L141-L158

Overall, spawning `esbuild` via its CLI seems far less than ideal. Instead, it should probably be executed via its nodejs api:

Example: 

```js
require('esbuild').buildSync({
  entryPoints: ['in.ts'],
  outfile: 'out.js',
})
```

Getting the installed version would be as easy as: `require('esbuild/package.json').version`

### Reproduction Steps

```
mkdir test
yarn init -2
...set up cdk project...
yarn add esbuild

yarn cdk deploy '*'
```

Prints: `esbuild cannot run locally. Switching to Docker bundling.`

A workaround is possible by installing `esbuild` globally, but this is far from ideal.

### What did you expect to happen?

`esbuild` should work properly

### What actually happened?

`esbuild cannot run locally. Switching to Docker bundling.`

### Environment

  - **CDK CLI Version  :** 1.90.1
  - **Framework Version:** 1.90.1
  - **Node.js Version:** 15.9.0
  - **OS               :** macOS
  - **Language (Version):** TypeScript 4.0.5

### Other

Originally added in https://github.com/aws/aws-cdk/pull/11289 by @jogold 

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by jogold**
Refactor how `esbuild` is detected and run: check for a local installation
first and fallback to a global installation.

For local bundling, if a local version is detected, run `esbuild` using
the right package manager.

In Docker, always run the globally installed version.

Support for PNPM could now easily be added in another PR.

Closes #13179

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by andreialecu** [comment]
@jogold I have a question about how this module should be used.

Say I have a `lambdas` folder underneath my main `cdk` project. Do I need to create a `package.json` and lockfile separately within it with dependencies? 

That seems to be the case with the current implementation, at least with yarn v2.

It would make it easier if that was not required, but instead it could take dependencies from the main cdk project's `package.json`. But currently that does not work.

**Comment by jogold** [comment]
@andreialecu

> Do I need to create a `package.json` and lockfile separately within it with dependencies?

No, you can have infra and runtime code next to each other with a **single** `package.json` and lock file for your whole project. See examples in https://github.com/jogold/cloudstructs and https://github.com/jogold/cdk-s3-thumbnail. What doesn't work for you when you try this?

**Comment by andreialecu** [comment]
@jogold I just tried this again, and if I delete the `package.json` and related files from the `lambdas` directory I'm getting:

```
 > create-auth-challenge.ts: error: Could not resolve "crypto-secure-random-digit" (mark it as external to exclude it from the bundle)
    3 │ import { randomDigits } from "crypto-secure-random-digit";
```

**Comment by jogold** [comment]
> @jogold I just tried this again, and if I delete the `package.json` and related files from the `lambdas` directory I'm getting:
> 
> ```
>  > create-auth-challenge.ts: error: Could not resolve "crypto-secure-random-digit" (mark it as external to exclude it from the bundle)
>     3 │ import { randomDigits } from "crypto-secure-random-digit";
> ```

Can you share a repo with this issue?

**Comment by andreialecu** [comment]
It seems that the problem is that this plugin would be needed:

https://github.com/yarnpkg/berry/tree/master/packages/esbuild-plugin-pnp

And unfortunately it only works with the build API, which cannot be used due to the aforementioned `jsii` issues.

I guess this will need to be mentioned as a caveat, that an intermediate `package.json` that uses plain `node_modules` needs to be used. Perhaps a CLI setting could be added to `esbuild` in order to specify plugins that way.

**Comment by jogold** [comment]
> It seems that the problem is that this plugin would be needed:

OK, does this means that this PR is currently useless because even if we correctly detect the presence of `esbuild`, it won't be able to resolve the modules when bundling in a PnP setup?

**Comment by andreialecu** [comment]
@jogold not necessarily, no.

The setup I described above works. It just needs an additional `package.json` and `node_modules` within the `lambdas` folder. 

Also I believe that https://github.com/evanw/esbuild/issues/884 might be a small change and I think it has the potential to be implemented in `esbuild`.

Alternatively, https://yarnpkg.com/features/protocols/#patch can be used to patch `esbuild` and force it to load the `pnp` plugin. It would be a minimal patch.

So this PR would still be a prerequisite for getting either of the two options above to work.

**Comment by andreialecu** [comment]
@jogold would it be feasible to allow overriding the builder, optionally, via a callback?

Here's an example of what I mean:

```ts
	import {build} from "esbuild";

    const cognitoEventsLambda = new lambda.NodejsFunction(this, id, {
      entry: path.resolve(__dirname, "cognito-lambdas", filename),
      handler: "handler",
      role: lambdaRole,
      builder: async (config) => {
         return await build({
           ...config,
           plugins: [pnpPlugin()],
         });
      }
    });
```

`builder` can be optional, and if specified, it is used instead of spawning the CLI here: https://github.com/aws/aws-cdk/pull/13257/files#diff-cd86fbd4f2bbefcbcffc2143adccabafa1debe5981edbcdfcc766b5a705fe770R154-R171

It would pass the `esbuild` config as a parameter, which the client side function can just use, or further modify.

This way it delegates to user land, so a peer dependency is no longer necessary, and everything would still work.

File: packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts [hunk 1]
```diff
@@ -57,11 +59,13 @@ export class Bundling implements cdk.BundlingOptions {
...
+  private readonly packageManager: PackageManager;
 
   constructor(private readonly props: BundlingProps) {
-    Bundling.runsLocally = Bundling.runsLocally
-      ?? getEsBuildVersion()?.startsWith(ESBUILD_VERSION)
-      ?? false;
+    this.packageManager = PackageManager.fromLockFile(props.depsLockFilePath);
+
+    Bundling.esbuildInstallation = Bundling.esbuildInstallation ?? EsbuildInstallation.detect();
+    const runsLocally = !!Bundling.esbuildInstallation?.version.startsWith(ESBUILD_VERSION);
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts:68]: maybe `esbuildInstallation?.isLocal`

File: packages/@aws-cdk/aws-lambda-nodejs/lib/esbuild-installation.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * An esbuild installation
+ */
+export abstract class EsbuildInstallation {
+  public static detect(): EsbuildInstallation | undefined {
+    try {
+      try {
+        // Check local version first
+        const version = getModuleVersion('esbuild');
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda-nodejs/lib/esbuild-installation.ts:12]: Why does this have to throw if the module doesn't exist? Can this be `tryGetModuleVersion`?

File: packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts (hunk 1 above)
Comment (jogold) [reply packages/@aws-cdk/aws-lambda-nodejs/lib/bundling.ts:68]: There too many locals... this `runsLocally` here means local bundling (vs Docker bundling). The `isLocal` of the `esbuildInstallation` means is locally installed (vs globally installed).

**Comment by jogold** [comment]
> @jogold would it be feasible to allow overriding the builder, optionally, via a callback?

We could maybe let the user implement `tryBundle()` (or a esbuild bundling options enriched version of it) here

https://github.com/aws/aws-cdk/blob/ad01099d5b8f835c3b87d7d20fd2dc1a5df2fd6f/packages/%40aws-cdk/aws-lambda-nodejs/lib/bundling.ts#L103

(but it will have to be `sync`)

**Comment by eladb** [comment]
@jogold sorry for the delay... Can you update this branch and let me know if it's ready for a review?

**Comment by eladb** [comment]
@jogold closing for now. reopen when ready to continue.

**Pull request closed without merging**